lib/web_ui/lib/src/engine/bitmap_canvas.dart

```diff
@@ -664,6 +664,201 @@ class BitmapCanvas extends EngineCanvas with SaveStackTracking {
     picture.recordingCanvas.apply(this);
   }
 
+  // Vertex shader transforms pixel space [Vertices.positions] to
+  // final clipSpace -1..1 coordinates with inverted Y Axis.
+  static const _vertexShaderTriangle = '''
+      #version 300 es
+      layout (location=0) in vec4 position;
+      layout (location=1) in vec4 color;
+      uniform vec4 u_scale;
+      uniform vec4 u_shift;
+      
+      out vec4 vColor;
+      void main() {
+        gl_Position = (position * u_scale) + u_shift;
+        vColor = color;
+      }''';
+  // This fragment shader enables Int32List of colors to be passed directly
+  // to gl context buffer for rendering by decoding RGBA8888.
+  static const _fragmentShaderTriangle = '''
+      #version 300 es
+      precision highp float;
+      in vec4 vColor;
+      out vec4 fragColor;
+      void main() {
+        fragColor = vec4(vColor[2], vColor[1], vColor[0], vColor[3]);
+      }''';
+
+  /// Draws vertices on a gl context.
+  ///
+  /// If both colors and textures is specified in paint data,
+  /// for [BlendMode.source] we skip colors and use textures,
+  /// for [BlendMode.dst] we only use colors and ignore textures.
+  /// We also skip paint shader when no texture is specified.
+  ///
+  /// If no colors or textures are specified, stroke hairlines with
+  /// [Paint.color].
+  ///
+  /// If colors is specified, convert colors to premultiplied (alpha) colors
+  /// and use a SkTriColorShader to render.
+  @override
+  void drawVertices(
```

Comment:
It seems like the `glCanvas` won't have the current transform applied, so the drawn vertices might be the wrong size when `devicePixelRatio != 1.0`.

Reply:
Added todo